### ISSUE TITLE
[FireMonkey] Display errors in tab console

### ISSUE THREAD

**Issue by pintassilgo** [erosman/support#307]
Currently, errors are displayed only in browser console.

Replacing

    script.js += sourceURL + encodeId + '.user.js';

by

    script.js = 'try{' + script.js + '}catch(e){console.error(e);}' + sourceURL + encodeId + '.user.js';

Should be enough.

The line, of course, will be the last of the code, but you can see the actual line in stack:

![image](https://user-images.githubusercontent.com/5483864/113190627-0d1b6100-924c-11eb-8609-9814db0e6966.png)

If you click the link in error, you go directly to the source in Debugger. This is much better than the current mode that only displays the error in Browser Console and in Browser Toolbox.

**Comment by erosman** [comment]
ATM on Nightly, in the **Developer Tools** it shows the correct line but shows the blob URL

![image](https://user-images.githubusercontent.com/6458876/113193632-6d72c880-9275-11eb-9598-38b6f16e85a6.png)


On click it shows

![image](https://user-images.githubusercontent.com/6458876/113193755-95622c00-9275-11eb-8f89-9d0a3cbfdc31.png)

Adding `try/catch` has no effect on above.

JavaScript Parse errors do end up in **Browser Cosole**

Let me try and see...

**Comment by pintassilgo** [comment]
> Adding try/catch has no effect on above.

It depends on the error. For me, if it's not a syntax error, tab console logs the error after the change I proposed.

So `teatae;` (undefined) will log in tab. But `tae tae;` (syntax error) is still exclusive to browser console, because JS parser can't know the end of `try` block..

I don't know if it's possible to log every error in tab console, but logging non-syntax errors is already an improvement.

**Comment by erosman** [comment]
I was referring to showing the script name. Even with the try/catch, it showed blob in Nightly.

I also tried `const a = def;` which is an undefined error.

![image](https://user-images.githubusercontent.com/6458876/113338297-2a305c80-933e-11eb-8156-d58ac4cdf1a2.png)

I tried `const abc def = 5;`

![image](https://user-images.githubusercontent.com/6458876/113338433-59df6480-933e-11eb-8ede-6e26843a7b39.png)


Have you tested it in Nightly 89?

**Comment by pintassilgo** [comment]
I use Developer Edition and runtime errors logs fine in tab console with the right URL.

![image](https://user-images.githubusercontent.com/5483864/113338973-5edded00-9319-11eb-842f-e9b861bce2e1.png)

Clicking in the URL goes directly to the source, but since it's a `try/catch` it selects the line inside catch instead of the line that triggered the error.

**Comment by erosman** [comment]
That doesn't happen in Nightly 89 with `try/catch` :man_shrugging:

**Comment by pintassilgo** [comment]
Fixed in commit above. Every image below is from **tab** console.

If the script has no issues...
![image](https://user-images.githubusercontent.com/5483864/113378003-6925d880-9365-11eb-95e8-b2ed33246617.png)

Runtime error, as GM and VM also can handle:
![image](https://user-images.githubusercontent.com/5483864/113378103-aee2a100-9365-11eb-8041-0e2cc62ae497.png)

Clicking the link, you go direct to source:
![image](https://user-images.githubusercontent.com/5483864/113378217-01bc5880-9366-11eb-900f-f0c73c771259.png)

Styntax Error, not even GM and VM can handle:
![image](https://user-images.githubusercontent.com/5483864/113378146-ce79c980-9365-11eb-9555-80ee3a4733bb.png)

Link works too:
![image](https://user-images.githubusercontent.com/5483864/113378289-2284ae00-9366-11eb-91dd-8f665926e818.png)

I just didn't tested for `@inject-into page`, I see no reason for a FireMonkey user to use that option, and since that way code is injected directly into page maybe we don't need to do anything to display errors.

**Comment by pintassilgo** [comment]
Fixed in https://github.com/pintassilgo/forked-firefox-extensions/commit/06d966a840fae415f84b3046e008db170d6f7428. Every image below is from **tab** console.

If the script has no issues...
![image](https://user-images.githubusercontent.com/5483864/113378003-6925d880-9365-11eb-95e8-b2ed33246617.png)

Runtime error, as GM and VM also can handle:
![image](https://user-images.githubusercontent.com/5483864/113378103-aee2a100-9365-11eb-8041-0e2cc62ae497.png)

Clicking the link, you go direct to source:
![image](https://user-images.githubusercontent.com/5483864/113378217-01bc5880-9366-11eb-900f-f0c73c771259.png)

Styntax Error, not even GM and VM can handle:
![image](https://user-images.githubusercontent.com/5483864/113378146-ce79c980-9365-11eb-9555-80ee3a4733bb.png)

Link works too:
![image](https://user-images.githubusercontent.com/5483864/113378289-2284ae00-9366-11eb-91dd-8f665926e818.png)

I just didn't tested for `@inject-into page`, I see no reason for a FireMonkey user to use that option, and since that way code is injected directly into page maybe we don't need to do anything to display errors.

**Comment by erosman** [comment]
`'unsafe-eval'` will cause the addon to be rejected. Using `eval()` on an untrusted code in privileged background script is a major security risk.

I had looked for ways to pass the errors at injection time but have not found a safe way.

**Comment by pintassilgo** [comment]
Just add `eval: window.eval` to `script.defineGlobals` and that's it.

Code submitted to AMO is reviewed, right? With that, I guess there's no way to exploit anything.

**Comment by pintassilgo** [comment]
Talking about this, why is `unsafeWindow` not defined in `script.defineGlobals`? It's a much more suited place than prepending in the source of each script.

**Comment by erosman** [comment]
Addon with 'unsafe-eval' in CSP will be rejected on AMO review.
_(PS. I have been an AMO add-on reviewer since 2014)_

Userscript that is going to be evaluated, is not packed with the addon so it is unknown code.

> Talking about this, why is unsafeWindow not defined in script.defineGlobals? It's a much more suited place than prepending in the source of each script.

I tried it in v1.47, but it broke it so I reverted it in v2.0. I am not sure why!!? It needs to be tested thoroughly to find the reason.

**Comment by pintassilgo** [comment]
> Userscript that is going to be evaluated, is not packed with the addon so it is unknown code.

With `eval: window.eval`, how could UserScripts benefit (exploit) from `unsafe-eval`?

> Addon with 'unsafe-eval' in CSP will be rejected on AMO review.

Right, I'll assume this as an fact, but I still wait for a reason specifically against what I'm proposing.

Anyway, just another thing so that I need to continue maintaining a fork. But that's OK, I'm used to it.

**Comment by erosman** [comment]
Example:

```js
// ==UserScript==
// @name          01
// @match         *://*.example.com/*
// ==/UserScript==

browser.storage.local.clear();
```

Above code has no effect in `userscript` context.

https://github.com/pintassilgo/forked-firefox-extensions/commit/06d966a840fae415f84b3046e008db170d6f7428#diff-59998e876bf86d9e080ffb9a116d490a2b384b6ee9632451510a4a8e53ff1077R200
```js
try{
  eval(`()=>{${script.js}\n}`);
} catch (e) { ... }
```

That is a simple example of the risk of using `eval()` in `background.js` which has full browser privilege.

**Comment by pintassilgo** [comment]
Since you edited it...

> That has no effect on background.js

And whats the point of `background.js`? The `eval` I added doesn't run nothing, it's an anonymous function with no references and that never is called. It may even be killed by gc. It's literally just a checker for compile-time errors.

> eval() is already available to userscripts since it is standard JavaScript.

Of course, but without `unsafe-eval` in manifest, UserScripts can't use `eval` in pages with CSP. With `unsafe-eval` they can. But then you set `eval: window.eval` and `unsafe-eval` has no effect for UserScripts. Also, Mozilla intends to stop pages from applying CSP to extensions' content scripts, so what's the point anyway?

**Comment by pintassilgo** [comment]
> Above code has no effect in userscript context. (but can be executed in content context)

Of course, I khow that.

> That is a simple example of the risk of using eval() in background.js which has full browser privilege

This is not a valid example. Did you noticed that code inside `eval` is never executed? It's inside an anonymous function that is not called nor assigned.

**Comment by erosman** [comment]
> Of course, but without unsafe-eval in manifest, UserScripts can't use eval in pages with CSP.

`'unsafe-eval'` in `manifest.json` applies to privileged code. User-scripts in `userScript` context are not privileged code.
User-scripts in `content` context are privileged code although it has lowest privilege comparing to extension code with browser privilege.

N.B. The `content` context is designed for addon's **own** code. It is a security risk that other manager inject the userscripts into  `content` context.

N.B. `eval()` anywhere in an add-on code will result in rejection of the add-on in 99.9% of the times. There is a very limited legitimate & acceptable use for `eval()` in any add-on. Not only `eval()` but there are other methods that evaluates untrusted code and they also result in rejection of the add-on due to the inherit security risks involved.

**Comment by pintassilgo** [comment]
Right, thanks for you attention, but I already know what you said and doesn't answer anything I said.

**Comment by erosman** [comment]
> Right, thanks for your attention, but I already know what you said and it doesn't answer anything I said.

So I misunderstood. What was the question?

**Comment by pintassilgo** [comment]
> I tried it in v1.47, but it broke it so I reverted it in v2.0. I am not sure why!!? It needs to be tested thoroughly to find the reason.

`unsafeWindow: window.wrappedJSObject` in `defineGlobals` is exactly the same as prepending `unsafeWindow= window.wrappedJSObject` in the UserScript, just more elegant. I saw the issue (#262) and saw v1.47 code, it was just your mistake... You had defined `unsafeWindow: window.wrappedJSObject` in `GM`, not in `defineGlobals`. So of course that broke `unsafeWindow`, people would have to use `GM.unsafeWindow`...

**Comment by erosman** [comment]
> unsafeWindow: window.wrappedJSObject in defineGlobals is exactly the same as prepending const unsafeWindow = window.wrappedJSObject in the UserScript, just more elegant. I saw the issue (#262) and saw v1.47 code, it was just your mistake... You had defined unsafeWindow: window.wrappedJSObject in GM, not in defineGlobals. So of course that broke unsafeWindow, people would have to use GM.unsafeWindow...

Can you run test and let me know? Or make a test script for me test?
Personally, I never use `unsafeWindow` so I don't know all the use cases to test it.

**Comment by pintassilgo** [comment]
> So I misunderstood. What was the question?

I think you simply didn't get that the `eval` to `background.js` doesn't run `anything` and doesn't expose anything.

1. The the `eval` in `background.js` doesn't execute nothing, it's **literally just a chacker for compile-time errors**, So it doesn't matter what people code in UserScript, they will never gain access to the higher context of `background.js`.

2. With or without `unsafe-eval`, `eval` in UserScripts run in the exact same context: global/this is the sandbox limited to what you chose to expose in `defineGlobals`. So it doesnt matter if you declare `unsafe-eval` or not, `eval` in UserScripts will never access internals like `browser`.

3. The **only** thing that `unsafe-eval` changes for UserScripts is that it bypass sites with CSP directives. Currently, without `unsafe-eval`, you can use `eval` in most pages, but not in those with CSP like GitHub and Twitter. Declaring `unsafe-eval` in manifest, you can `eval` in pages with CSP just as you do in pages without CSP. Nothing more than that. No access to privileged code is given.

4, Even if **3** is a problem for you*, you can just set `eval: window.eval` and then UserScripts will be blocked again to use `eval` in pages with CSP. With that, there's absolutely no difference at the end, `unsafe-eval` won't affect users, won't expose anything, won't allow higher privilege to unknown code.

**Comment by erosman** [comment]
As I added to the previous post ...

### `eval()` or `unsafe-eval` are not allowed in add-ons on AMO.

Therefore, all discussion are purely academic.

**Comment by pintassilgo** [comment]
So that's it. You can't give a reason.

As you said, "*there is a very limited legitimate & acceptable use for eval() in any add-on*" and I'm trying to prove you this is the case, but you're not understanding my point or simply don't want to think about it.

**Comment by erosman** [comment]
It doesn't matter if someone can prove it or not. The AMO policy does not allow it. Therefore, there is no point in including it in the add-on. Furthermore, it is not an imperative feature.

Also, the following gives no error

```js
try { 
  eval(`()=>{let a = abc;}`);
 } catch(error) { console.error(error.message); }
```

**Comment by pintassilgo** [comment]
> Personally, I never use unsafeWindow so I don't know all the use cases to test it.

I tried to explain about this yesterday in the other issue. The best you can get is there:
https://blog.mozilla.org/addons/2014/04/10/changes-to-unsafewindow-for-the-add-on-sdk/

`unsafeWindow` was introduced by Greasemonkey in 2005 to limit the access pages can have from browser internals when there's an UserScript running.

Before that, you could expose anything to the web by doing a simple `window.GM = GM`. After GM introduced `unsafeWindow`, using `window` in userscripts no longer refers to the real `window` object in the page, but instead to something more restricted. You still had `window.document`, `window.Array` and like, plus the GM APIs, but only those cherry-picked objects, no sensitive ones like `Components` (you're also old in Fx, so I'll assume you know about `Components`). From then on, UserScripts writers would have to use `unsafeWindow` to get the true `window` object from the page, . So it wasn't a real sandbox, just made harder for UserScript writers who didn't understand about Fx internals to open security holes by mistake.

The true sandbox came when not even `unsafeWindow` refers anymore to the true `window` in page. That's why Mozilla created `cloneInto`/`exportFunction`, as the only channel to expose objects from privileged context to the page.

Since the real sandbox was implemented, the difference between `window` and `unsafeWindow` (which is nothing more than an alias to `window.wrappedJSObject`) is that with `unsafeWindow` you can create global variables to the page **if they have primitive values** (number, string).

> Can you run test and let me know? Or make a test script for me test?

Inside `api.js` (`api_script` in manifest), `window` is exactly the same `window` exposed in userScripts context. You can confirm that easily: just set `win: window` in `script.defineGlobals`, then create an UserScript with `console.log(win == window)`. You'll get `true`.

So you should define `unsafeWindow: window.wrappedJSObject` in `defineGlobals`. This is not going to change anything, it's only more suitable and elegant.

**Comment by erosman** [comment]
> So you should define unsafeWindow: window.wrappedJSObject in defineGlobals. This is not going to change anything, it's only more suitable and elegant.

I agree that it is better to define it there.

I need to test it to make sure there are no problems for the users.
Since I dont use `unsafeWindow` in my own userscripts, if someone can make a test userscript with various uses of `unsafeWindow`, I will test it out.

**Comment by pintassilgo** [comment]
> PS. the following gives no error

You still don't understand...

Paste that in console and neither you will see an error:

`()=>{eros_man}`

That's because **the function is not executed**. That's why I'm saying from the start this has no security implications.

When you define a function but don't call it, the code can have infinite runtime erros like undefined variables and browser won't complain. **Run**-time errors, so they are only triggered when the function is called.

When you declare a function without calling it, only `compile-time errors` are triggered, like syntax errors. This is a compile-time error:

`()=>{eros{_man}`

That's why I added **two** `console.error` in `background.js`. One for compile-time errors and the other for runtime errors. `try catch` can only catch runtime errors, because `syntax error` prevents the JS engine to know where the `try` block ends.

**Comment by pintassilgo** [comment]
> I just tested it and unsafeWindow doesn't work this way.

Example?

**Comment by erosman** [comment]
> You still don't understand...

I see... so the `eval()` in your example only reveals the syntax errors.
Nonetheless, it is not allowed on AMO.

> Example?

I tested the example in #262

```js
let confirm = new window.Function('return true');
unsafeWindow.confirm = confirm;
```

**Comment by pintassilgo** [comment]
> I tested the example in #262

It works.

**Comment by erosman** [comment]
I didnt work for me :man_shrugging: 


```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <title>Test Page</title>
  </head>
  <body>

<button id="confirm" type="button">Confirm</button>

<script type="text/javascript">
  
  document.querySelector('#confirm').addEventListener('click', () => {
    console.log(confirm('This is page confirm'));;
  });
</script>

</body>
</html>
```

```js
let confirm = new window.Function('return true');
unsafeWindow.confirm = confirm;
```

When I tested with `unsafeWindow` in `script.defineGlobals` it doesn't work, but works when `unsafeWindow` is set as it is now in `background.js`.

**Comment by pintassilgo** [comment]
You're right about `unsafeWindow`. I hadn't realized that I added `unsafeWindow:  window.wrappedJSObject` in `api.js` but forgot to remove `unsafeWindow` definition in `background.js`, that's why it was working. 🤦‍♂️

`api_script` `window` is not the same as `window` in userScript context. There's no way to get, from `api.js`, the same object as `window.wrappedJSObject` in userScript context.

**Comment by erosman** [comment]
Monaco Editor plus ESLint should pick up most of the errors in FireMonkey 3.0.
